### ISSUE TITLE
Unable to remote into windows 10 machine using Tailscale intermittently.

### ISSUE THREAD

**Issue by darshinimashar** [tailscale/tailscale#1448]
Try to login to windows 10 machine from macOS using Tailscale. 

Windows Defender Firewall is turned off, Allow incoming portion is selected, Run unattended enabled.

Error on destination machine:

2021-03-04 11:41:41.9635918 -0500 EST: [lHW9R] - ConsumeMessageInitiation: handshake replay @ [64 0 0 0 96 65 13 207 36 0 0 0]
2021-03-04 11:41:41.9635918 -0500 EST: Received invalid initiation message from 9ed70c933fb8c443af7369875ad2c1e565c88cbf9036e3fe807ece114865f31e.disco.tailscale:12345

<img src="https://frontapp.com/assets/img/favicons/favicon-32x32.png" height="16" width="16" alt="Front logo" /> [Front conversations](https://app.frontapp.com/open/top_2vcb5)

**Comment by bradfitz** [comment]
Source was:
[lHW9R] is 100.108.150.50 (logid a267ec71300097933251c8f5cc9336cc0ea12b2379b085920ee3391c343ac4d2)

What was the destination machine?

**Comment by darshinimashar** [comment]
destination is 2021-03-04 11:52:01.7217496 -0500 EST: netmap: self: [7XbOw] auth=machine-authorized u=darshimashar18@gmail.com [100.67.68.57/32]

**Comment by darshinimashar** [comment]
Also, the machine is using node sharing to my tailscale network. I am sharing 100.67.68.57 machine from darshimashar18@gmail.com to another domain.

**Comment by bradfitz** [comment]
Well, the wireguard handshake succeeded, as we got to the ACL "Accept" lines:

```
2021-03-04 11:38:03.1824488 -0500 -0500: magicsock: got updated network map; 5 peers (5 with discokey)
2021-03-04 11:41:41.9625918 -0500 -0500: magicsock: disco: d:5600b0b63e6b50b2<-d:9ed70c933fb8c443 ([lHW9R], 192.168.0.98:41641)  got ping tx=239a1a4ff3b5
2021-03-04 11:41:41.9625918 -0500 -0500: [lHW9R] - Received handshake initiation
2021-03-04 11:41:41.9625918 -0500 -0500: [lHW9R] - Sending handshake response
2021-03-04 11:41:41.9635918 -0500 -0500: magicsock: disco: d:5600b0b63e6b50b2<-d:9ed70c933fb8c443 ([lHW9R], derp-1)  got call-me-maybe, 8 endpoints
2021-03-04 11:41:41.9635918 -0500 -0500: [lHW9R] - ConsumeMessageInitiation: handshake replay @ [64 0 0 0 96 65 13 207 36 0 0 0]
2021-03-04 11:41:41.9635918 -0500 -0500: Received invalid initiation message from 9ed70c933fb8c443af7369875ad2c1e565c88cbf9036e3fe807ece114865f31e.disco.tailscale:12345
2021-03-04 11:41:41.9635918 -0500 -0500: magicsock: disco: d:5600b0b63e6b50b2->d:9ed70c933fb8c443 ([lHW9R], 97.108.50.244:1056) sent ping tx=1da88b91e8e9
2021-03-04 11:41:41.9635918 -0500 -0500: magicsock: disco: d:5600b0b63e6b50b2->d:9ed70c933fb8c443 ([lHW9R], 192.168.0.98:41641) sent ping tx=d2467a743d37
2021-03-04 11:41:41.975599 -0500 -0500: magicsock: disco: d:5600b0b63e6b50b2<-d:9ed70c933fb8c443 ([lHW9R], 192.168.0.98:41641)  got pong tx=d2467a743d37 latency=13ms pong.src=192.168.0.75:41641
2021-03-04 11:41:41.975599 -0500 -0500: Accept: TCP{100.108.150.50:61849 > 100.67.68.57:3389} 64 tcp ok
2021-03-04 11:41:42.1301836 -0500 -0500: logtail: dialed "log.tailscale.io:443" in 155ms
2021-03-04 11:41:42.7552877 -0500 -0500: Accept: TCP{100.108.150.50:61849 > 100.67.68.57:3389} 64 tcp ok
2021-03-04 11:41:43.5720793 -0500 -0500: v1.4.5-tb89c65304-gf09ac28c2 peers: 808/248 10496/1568
2021-03-04 11:41:43.7600184 -0500 -0500: Accept: TCP{100.108.150.50:61849 > 100.67.68.57:3389} 64 tcp ok
2021-03-04 11:41:45.571637 -0500 -0500: v1.4.5-tb89c65304-gf09ac28c2 peers: 808/248 10592/1568
2021-03-04 11:41:47.5711348 -0500 -0500: v1.4.5-tb89c65304-gf09ac28c2 peers: 808/248 10784/1568
2021-03-04 11:41:51.9764498 -0500 -0500: [lHW9R] - Sending keepalive packet
2021-03-04 11:41:53.5722861 -0500 -0500: v1.4.5-tb89c65304-gf09ac28c2 peers: 808/248 10784/1600
2021-03-04 11:42:01.5594371 -0500 -0500: control: mapRoutine: netmap received: state:synchronized
2021-03-04 11:42:01.5604345 -0500 -0500: control: sendStatus: mapRoutine-got-netmap: state:synchronized
2021-03-04 11:42:01.5604345 -0500 -0500: netmap diff:
2021-03-04 11:42:01.5604345 -0500 -0500: - [nvSKT] d:604c4a4208435134 D9 100.76.112.103  :   97.108.50.244:1051     192.168.0.99:41641
2021-03-04 11:42:01.5604345 -0500 -0500: + [nvSKT] d:604c4a4208435134 D1 100.76.112.103  :   97.108.50.244:1051     192.168.0.99:41641
2021-03-04 11:42:01.5614345 -0500 -0500: magicsock: got updated network map; 5 peers (5 with discokey)
2021-03-04 11:42:01.9785367 -0500 -0500: [lHW9R] - Sending keepalive packet
2021-03-04 11:42:03.5721235 -0500 -0500: v1.4.5-tb89c65304-gf09ac28c2 peers: 808/248 10784/1632
2021-03-04 11:42:42.0101736 -0500 -0500: magicsock: disco: send, starting discovery for [lHW9R] (d:9ed70c933fb8c443)
2021-03-04 11:42:42.0121702 -0500 -0500: magicsock: want call-me-maybe but endpoints stale; restunning
2021-03-04 11:42:42.0121702 -0500 -0500: magicsock: starting endpoint update (refresh-for-peering)
2021-03-04 11:42:42.0121702 -0500 -0500: magicsock: disco: d:5600b0b63e6b50b2->d:9ed70c933fb8c443 ([lHW9R], 97.108.50.244:1056) sent ping tx=7443f3ab0440
2021-03-04 11:42:42.0161719 -0500 -0500: magicsock: disco: d:5600b0b63e6b50b2<-d:9ed70c933fb8c443 ([lHW9R], derp-1)  got call-me-maybe, 8 endpoints
2021-03-04 11:42:42.0161719 -0500 -0500: magicsock: disco: d:5600b0b63e6b50b2->d:9ed70c933fb8c443 ([lHW9R], 97.108.50.244:1056) sent ping tx=0502b83ba0c7
2021-03-04 11:42:42.0161719 -0500 -0500: magicsock: disco: d:5600b0b63e6b50b2->d:9ed70c933fb8c443 ([lHW9R], 192.168.0.98:41641) sent ping tx=ab8c9b64931d
2021-03-04 11:42:42.0266109 -0500 -0500: magicsock: disco: d:5600b0b63e6b50b2<-d:9ed70c933fb8c443 ([lHW9R], 192.168.0.98:41641)  got pong tx=ab8c9b64931d latency=10ms pong.src=192.168.0.75:41641
2021-03-04 11:42:42.7835616 -0500 -0500: netcheck: probePortMapServices: me 192.168.18.1 -> gw 192.168.0.1
2021-03-04 11:42:42.9093043 -0500 -0500: netcheck: report: udp=true v6=false mapvarydest=false hair=false portmap= v4a=97.108.50.244:41641 derp=1 derpdist=1v4:44ms,2v4:118ms,3v4:279ms,4v4:129ms,5v4:316ms,6v4:246ms,7v4:206ms,8v4:121ms,9v4:64ms,10v4:90ms
2021-03-04 11:42:43.0199293 -0500 -0500: magicsock: STUN done; sending call-me-maybe to d:9ed70c933fb8c443 [lHW9R]
2021-03-04 11:42:43.0204733 -0500 -0500: magicsock: disco: d:5600b0b63e6b50b2->d:9ed70c933fb8c443 ([lHW9R], derp-1) sent call-me-maybe
2021-03-04 11:42:46.542371 -0500 -0500: Accept: TCP{100.108.150.50:61863 > 100.67.68.57:3389} 64 tcp ok
2021-03-04 11:42:47.5744862 -0500 -0500: v1.4.5-tb89c65304-gf09ac28c2 peers: 808/248 10880/1632
2021-03-04 11:42:47.5795945 -0500 -0500: Accept: TCP{100.108.150.50:61863 > 100.67.68.57:3389} 64 tcp ok
2021-03-04 11:42:48.1645792 -0500 -0500: Accept: TCP{100.108.150.50:61863 > 100.67.68.57:3389} 64 tcp ok
2021-03-04 11:42:49.5725658 -0500 -0500: v1.4.5-tb89c65304-gf09ac28c2 peers: 808/248 11168/1632
2021-03-04 11:42:51.5702335 -0500 -0500: v1.4.5-tb89c65304-gf09ac28c2 peers: 808/248 11360/1632
2021-03-04 11:42:53.5717162 -0500 -0500: peer keys: [Fk5XX] [lHW9R]
2021-03-04 11:42:53.5727127 -0500 -0500: v1.4.5-tb89c65304-gf09ac28c2 peers: 808/248 11456/1632
2021-03-04 11:42:56.5408577 -0500 -0500: [lHW9R] - Sending keepalive packet
2021-03-04 11:42:57.5707823 -0500 -0500: v1.4.5-tb89c65304-gf09ac28c2 peers: 808/248 11456/1664
2021-03-04 11:42:57.5773296 -0500 -0500: Accept: TCP{100.108.150.50:61863 > 100.67.68.57:3389} 64 tcp ok
2021-03-04 11:42:59.5714687 -0500 -0500: v1.4.5-tb89c65304-gf09ac28c2 peers: 808/248 11552/1664
```

So we delivered the TCP SYN to Windows. Then what happened?

I don't see any `setFirewalPrivate` errors. All the "router: firewall" programming lines look happy too.

Could you grab a Wireshark packet capture on the Tailscale device and see if they're being injected into Windows fine?

**Comment by darshinimashar** [comment]
Sure

**Comment by darshinimashar** [comment]
![image](https://user-images.githubusercontent.com/75628398/110013097-37cad600-7cef-11eb-91d7-8a0903e82981.png)

**Comment by bradfitz** [comment]
So, we got the Wireguard, decrypted it, and gave the TCP SYN to wintun and the OS got it.

What did Windows do to it?

/cc @peske @danderson

**Comment by bradfitz** [comment]
Is there some RDP-specific firewall eating these 3389 packets?

**Comment by darshinimashar** [comment]
I will run capture at the windows end and share it with you.

—
Darshini Mashar | Customer Engineer | Tailscale | @tailscale

--- original message ---
On March 4, 2021, 1:49 PM EST notifications@github.com wrote:



Is there some RDP-specific firewall eating these 3389 packets?

—

You are receiving this because you were assigned.

Reply to this email directly, view it on GitHub, or unsubscribe.
--- end of original message ---

**Comment by bradfitz** [comment]
Oh, I thought that was the Windows end. Now I see "utun" and understand it was the Mac end.

**Comment by darshinimashar** [comment]
After rebooting Windows 10 machine, it fixed the issue with **100.67.68.57**.

here is the Wireshark capture from windows server 2019:  **100.76.112.103**
[test_windows.pcapng.zip](https://github.com/tailscale/tailscale/files/6086192/test_windows.pcapng.zip)
Pcap from

**Comment by darshinimashar** [comment]
@bradfitz here is the capture from both ends. 

[mac_capture.pcapng.zip](https://github.com/tailscale/tailscale/files/6086746/mac_capture.pcapng.zip)
[test_windows-68.57.pcapng.zip](https://github.com/tailscale/tailscale/files/6086747/test_windows-68.57.pcapng.zip)

**Comment by bradfitz** [comment]
We can reproduce this now and see what's going on between Tailscale and Windows. Fix on its way.

@danderson will probably be doing the fix, so assigned to him.

**Comment by danderson** [comment]
An unstable build with this fix is available at https://pkgs.tailscale.com/unstable/tailscale-ipn-setup-1.5.221.exe . We're going to cut a 1.4.6 patch release tomorrow, and this fix will be included in 1.6.0.

Writing up a root cause and how we figured it out, for posterity...

**Comment by danderson** [comment]
Okay, so the root cause of this problem was a bug in the routing table synchronization logic on Windows, coupled by some surprising behavior on the part of Windows.

When we create a Wintun adapter on Windows and add an IP address to it, Windows automatically creates a "system" route for the interface IP, pointing to the interface:

```
> netsh interface ipv4 show route
Publish  Type      Met  Prefix                    Idx  Gateway/Interface Name
-------  --------  ---  ------------------------  ---  ------------------------
...
No       System    256  100.81.72.5/32              5  Tailscale
...
```

Then, when we add routes to the rest of the tailnet, those show up as further "manual" routes:

```
> netsh interface ipv4 show route
Publish  Type      Met  Prefix                    Idx  Gateway/Interface Name
-------  --------  ---  ------------------------  ---  ------------------------
No       Manual    0    100.65.206.27/32            5  100.81.72.5
No       System    256  100.81.72.5/32              5  Tailscale
No       Manual    0    100.100.100.100/32          5  100.81.72.5
```

In the `wgengine/router` package, we implement address and route synchronization: we dump the current state of the network stack, diff it with our desired state (derived from Tailscale's network map), and issue the appropriate address/route add/delete commands to resync the state. We do this because the alternative would be to flush all Tailscale routes and reprogram them from scratch, which breaks active sockets and is generally quite disruptive - not great when devices potentially come and go frequently.

The issue ended up being that our route synchronization logic would find the Windows-added system route, see that it's not part of the expected route state (since we don't want to add it explicitly, Windows does it for us), and deleted it.

Deleting that system route breaks all incoming traffic through Tailscale, in a very mysterious way: packets flow successfully all the way through Tailscale from the client to the Windows machine, through the Wintun interface (we see packet counters increment), and then vanish from the network stack without a trace. Once we were able to reproduce the issue reliably, we traced the incoming packets and tested multiple hypotheses (misconfigured firewall, wrong interface profile, race condition in our code that configures WireGuard, ...), and eventually zeroed in on a diff in `netsh`'s route dump (slightly more detailed than `netstat -r`), which showed the system route present on a working Windows setup, and missing on a broken one.

Why is the problem intermittent? Because it so happens that there is a fair amount of delay (seconds+) between us adding an address to an interface, and the system route appearing in Windows's route dump APIs. So, most of the time, when Tailscale first starts up, it adds interface IPs (which implicitly creates the important system route), but then the route synchronization logic _doesn't_ see that system route when it does its route diff to calculate required changes. And so, it sets up the network stack correctly, and packets flow. Later on, if a netmap update happens that causes an OS network reconfiguration (e.g. adding a new device to your tailnet), _then_ the route sync logic finds and deletes the system route, and breaks incoming connectivity.

The problem is exacerbated by caches in the Windows network stack: as several of you noted, connecting out from an afflicted Windows machine to another device in your tailnet also fixes connectivity in reverse, back into the Windows machine - but only for that one peer that the Windows machine talked to. We haven't worked out all the details, but it seems that this is because outbound traffic creates some route lookup cache entries, which return traffic can then utilize to be correctly processed by the Windows machine. So, for some amount of time after doing the "reverse ping", connectivity works fine between that pair of machines, but still not to anywhere else. And if you let that link go idle long enough, the route cache entries expire, and you go back into the broken state.

We believe this explains all combinations of symptoms we've encountered ourselves, and heard from users to date:
 - "New users on my tailnet can't RDP to my Windows machine": in fact, theoretically _nobody_ can RDP in, but existing users are probably getting by on the route lookup cache, whereas new devices don't have that luxury.
 - "I went home for the weekend and when I came back on Monday my Windows machine was unreachable": it probably got into the broken state at some point on Friday, and then the weekend idle time was enough to expire route caches.
 - "Nobody can connect until I restart Tailscale on the Windows machine": restarting tailscale flushes all the interface configuration, so as long as there is only 1 round of network configuration during startup (the most common case), it's very likely (but not certain) that the system route propagation race will happen and allow us to correctly configure the OS network once, which fixes connectivity until the next big netmap change (e.g. device added/removed).
 - "If I ping from Windows I can RDP in": yup, because outbound traffic sets up those cache entries that allow the inbound traffic to make it despite the missing routing state.

The unstable build I linked above contains the bugfix, which explicitly looks for and preserves the system routes created by Windows when doing config synchronization. As I said above, the plan is to cut a 1.4 patch release (with a signed Windows build - our unstable builds aren't signed), and the fix will roll into the upcoming 1.6.0 as well.

Thanks for bearing with us while we tracked this down. I know it's been a frustrating time for some of you dealing with this, so we're anxious to have y'all verify the fix!
